frontend/model/contracts/group.js

```diff
@@ -312,32 +312,138 @@ sbp('chelonia/defineContract', {
     groupMincomeCurrency (state, getters) {
       return getters.groupSettings.mincomeCurrency
     },
+    // Oldest period key first.
+    groupSortedPeriodKeys (state, getters) {
+      // The .sort() call might be only necessary in older browser which don't maintain object key ordering.
+      // A comparator function isn't required for now since our keys are ISO strings.
+      return Object.keys(getters.currentGroupState.paymentsByPeriod ?? {}).sort()
+    },
+    // Returns either the known period stamp for the given date,
+    // or the predicted one according to the period length.
+    // May return 'undefined', in which case the caller should check archived data.
     periodStampGivenDate (state, getters) {
       return (recentDate: string | Date) => {
-        if (typeof recentDate !== 'string') {
-          recentDate = recentDate.toISOString()
-        }
+        if (typeof recentDate !== 'string') recentDate = recentDate.toISOString()
+        if (!isPeriodStamp(recentDate)) throw new TypeError('must be date or isostring')
         const { distributionDate, distributionPeriodLength } = getters.groupSettings
-
-        if (!distributionDate) return null
-
-        return periodStampGivenDate({
-          recentDate,
-          periodStart: distributionDate,
-          periodLength: distributionPeriodLength
-        })
+        if (!distributionDate) return
+        const sortedPeriodKeys = getters.groupSortedPeriodKeys
+        // Maybe the distribution date has just been updated,
+        // but the corresponding period has not started or is not stored yet.
+        // So we try to use the distribution date first.
+        if (recentDate >= distributionDate) {
+          // Only extrapolate one period length in the future.
+          const extrapolatedDistributionDate = addTimeToDate(
+            distributionDate, distributionPeriodLength
+          ).toISOString()
+          if (recentDate >= extrapolatedDistributionDate) {
+            return dateToPeriodStamp(extrapolatedDistributionDate)
+          }
+          return dateToPeriodStamp(distributionDate)
```

Comment:
Maybe you're right, but I haven't seen any use case where extrapolating more than one period in the future would be actually useful. I thought it didn't make much sense.

Reply:
This block is only called when `recentDate >= distributionDate`, and I don't think `distributionDate` can depend on the user's clock.
Therfeore in case the clock incorrectly thinks it's already several period lengths in the future past the distribution date, so that this block gets called, there just won't be any matching payment to be displayed.

If the user has their clock accidentally incorrectly set a year in the past, so that `recentDate` is one year lower than expected, then this block won't be called. The sorted periods will be searched instead, therefore I think the user should still be able to browse any stored payment period up to the distribution date.

Reply:
Updated to handle future dates using the period length in a loop like before (re-using `periodStampGivenDate` in `time.js`)